### DISCUSSION TITLE
Non-zero velocity at a mesh boundary

### DISCUSSION THREAD

**Discussion by pycam70** [pyvista/pyvista#1869] [Q&A]
Hi !

I am a master student specialized in fluid mechanics and I would like some help for my team project which consists of creating a work-flow for CFD data-processing from Openfoam via Pyvista.

Actually we are trying to calculate the wall shear stress (https://www.cfd-online.com/Wiki/Wall_shear_stress) at the boundary of the mesh.

As an easy first step, we are trying to plot the velocity along the boundary of our mesh. 
In  fluid mechanics, the no-slip condition (https://en.wikipedia.org/wiki/No-slip_condition) at the boundary should be checked (the three components shoud be equal to zero for a viscous fluids). However, when we plot the velocity, the X-component is not equal to zero even if we choose cell data or point data. 

Here is the code to create a line at the bottom boundary : 
`a = [0, mesh_point.bounds[2], 0]`
`b = [mesh_point.bounds[1]-0.085, mesh_point.bounds[2], 0]`

`line1 = pv.Line(a, b)`

`p10 = pv.Plotter()`
`p10.add_mesh(mesh_point, style="wireframe", color='white')`
`p10.add_mesh(line1, color="r", line_width=10)`
`p10.view_xy()`
`p10.add_axes()`
`p10.show()`

![image](https://user-images.githubusercontent.com/92042715/143682892-f17236c0-7ed1-491a-aa24-732790e2652a.png)
![image](https://user-images.githubusercontent.com/92042715/143683153-5959151d-66ea-44af-a44b-6523b5211947.png)

Then, we plot the velocity over the line: 

With point data: 
`mesh_point.plot_over_line(a, b, resolution=500, scalars='U')`

![image](https://user-images.githubusercontent.com/92042715/143683029-97d65bb6-eabf-40d4-abf2-7c5c9c9900e5.png)

With cell data (mesh_cell contains the cell data): 
`mesh_cell.plot_over_line(a, b, resolution=500, scalars='U')`

![image](https://user-images.githubusercontent.com/92042715/143683060-f15eae40-d213-4791-915e-58c3d733e3d7.png)

How to access the velocity data at the boundary (which shoud be null) ?
Could someone help us ?

**Comment by MatthewFlamm** [comment]
In OpenFOAM are you using a turbulence model?  Are you accessing the boundary mesh or the interior cell zone mesh in this example?  

If you are using a turbulence model with a wall model, the velocity will not appear to be zero at the wall in the interior of the domain.  However, if you use the boundary mesh itself, you _may_ get a zero velocity possibly depending on the wall model and turbulence model.  Beware that you need to use the wall model equations to get the wall shear stress correct as well.

**Reply by pycam70** [reply]
Thank you for your response !

We are using a turbulence model (k-epsilon) and wall model equation of the boundary layer theory which respect the no-slip condition. 
How can we use the boundary mesh itself to access the velocity data at the boundary ? 
I think that it was possible thanks to a line created at the boundary but it seems to not be a good idea because the velocity is not null.

**Reply by MatthewFlamm** [reply]
How did you get `mesh_point` in your example?  Probably something like `mesh_point = mesh[0]`?  If so, you should look at the other entries in `mesh`.

**Reply by MatthewFlamm** [reply]
> I think that it was possible thanks to a line created at the boundary but it seems to not be a good idea because the velocity is not null.

This is because you are using the interior domain mesh that has cell centered values.  pyvista and vtk know nothing about the wall model in use in openFOAM and therefore extrapolate incorrectly to the surface for wall parallel directions.

**Reply by pycam70** [reply]
Yes, but `mesh` only has one component in my case : the unstructuredGrid that contains the internal mesh.

**Reply by MatthewFlamm** [reply]
I have a local branch that exposes the API for reading in the 'patches' values in `pyvista.OpenFOAMReader`.  I will try to push this soon so you can test it.  You should look into whether you can get OpenFOAM to directly output the wall shear stress onto the patch data of the boundaries, this will be the most accurate.

**Reply by pycam70** [reply]
Great ! Keep me in touch. 

Yes, it is easy to get the wall shear stress in OpenFOAM, but the aim of our project is to create a post-processing workflow that wouldn't affect the data obtained from OpenFOAM.

**Reply by MatthewFlamm** [reply]
Can you test with #1896?  You would do something like:

```py
import pyvista
reader = pyvista.OpenFOAMReader(filename)
reader.enable_all_patch_arrays()
# set active time point if needed
# reader.set_active_time_value(5.0)
mesh = reader.read()
print(mesh.keys())  # should have a second multiblock containing the boundaries
```

**Reply by pycam70** [reply]
Thanks for your code ! 
Unfortunetly, it doesn't work : It shows me " 'OpenFOAMReader' object has no attribute 'enable_all_patch_arrays' ".
What argument should I put ? I have tried to write the path to the boundary file ( ... \constant\polyMesh) but it doesn't work.

**Reply by MatthewFlamm** [reply]
> Unfortunetly, it doesn't work :

You will need to install pyvista from the pull request in #1896 (or the branch).  If you are using pip (I would recommend a separate virtual environment for this exercise), you would do `pip install git+https://github.com/pyvista/pyvista.git@refs/pull/1896/head`.

**Reply by pycam70** [reply]
Thank you very much for your help ! It works :)
I will soon study this deeper but it seems to be working well : All the boundary conditions are respected and I can get the "zero velocity" at the mesh boundary.

**Reply by MatthewFlamm** [reply]
This functionality is available starting in version 0.33.0
